zone/cheat_manager.cpp

```diff
@@ -0,0 +1,382 @@
+					"/MQGate used at x [{:.2f}] y [{:.2f}] z [{:.2f}]",
+					position1.x,
+					position1.y,
+					position1.z
+				);
+				database.SetMQDetectionFlag(
+					m_target->AccountName(),
+					m_target->GetName(),
+					message.c_str(),
+					zone->GetShortName()
+				);
+				LogCheat(message);
+			}
+			break;
+		case MQGhost:
+			if (RuleB(Zone, EnableMQGhostDetector) &&
+				((m_target->Admin() < RuleI(Zone, MQGhostExemptStatus) || (RuleI(Zone, MQGhostExemptStatus)) == -1))) {
+				database.SetMQDetectionFlag(
+					m_target->AccountName(),
+					m_target->GetName(),
+					"/MQGhost",
+					zone->GetShortName());
+				LogCheat("/MQGhost");
+			}
+			break;
+		case MQFastMem:
+			if (RuleB(Zone, EnableMQFastMemDetector) &&
+				((m_target->Admin() < RuleI(Zone, MQFastMemExemptStatus) ||
+				  (RuleI(Zone, MQFastMemExemptStatus)) == -1))) {
+				std::string message = fmt::format(
+					"/MQFastMem used at x [{:.2f}] y [{:.2f}] z [{:.2f}]",
+					position1.x,
+					position1.y,
+					position1.z
+				);
+				database.SetMQDetectionFlag(
+					m_target->AccountName(),
+					m_target->GetName(),
+					message.c_str(),
+					zone->GetShortName()
+				);
+				LogCheat(message);
+			}
+			break;
+		default:
+			std::string message = fmt::format(
+				"Unhandled HackerDetection flag with location from x [{:.2f}] y [{:.2f}] z [{:.2f}]",
+				position1.x,
+				position1.y,
+				position1.z
+			);
+			database.SetMQDetectionFlag(
+				m_target->AccountName(),
+				m_target->GetName(),
+				message.c_str(),
+				zone->GetShortName()
+			);
+			LogCheat(message);
+			break;
+	}
+}
+
+void CheatManager::MovementCheck(glm::vec3 updated_position)
+{
+	if (m_time_since_last_movement_history.GetRemainingTime() == 0) {
+		CheatDetected(MQWarp, updated_position);
+	}    // someone is spoofing.
+
+	float  dist     = DistanceNoZ(m_target->GetPosition(), updated_position);
+	uint32 cur_time = Timer::GetCurrentTime();
+	if (dist == 0) {
+		if (m_distance_since_last_position_check > 0.0f) {
+			MovementCheck(0);
+		}
+		else {
+			m_time_since_last_position_check = cur_time;
+			m_cheat_detect_moved             = false;
+		}
+	}
+	else {
+		m_distance_since_last_position_check += dist;
+		m_cheat_detect_moved = true;
+		if (m_time_since_last_position_check == 0) {
+			m_time_since_last_position_check = cur_time;
+		}
+		else {
+			MovementCheck(2500);
+		}
+	}
+}
+
+void CheatManager::MovementCheck(uint32 time_between_checks)
+{
+	uint32 cur_time = Timer::GetCurrentTime();
+	if ((cur_time - m_time_since_last_position_check) > time_between_checks) {
+		float speed =
+				  (m_distance_since_last_position_check * 100) / (float) (cur_time - m_time_since_last_position_check);
+		int   runs  = m_target->GetRunspeed() / std::min(RuleR(Zone, MQWarpDetectionDistanceFactor), 1.0f);
+		if (speed > runs) {
+			if (!m_target->GetGMSpeed() && (runs >= m_target->GetBaseRunspeed() || (speed >
+																					(m_target->GetBaseRunspeed() /
```

Comment:
Can you make this line into a meaningful variable or break these checks down into bool variables that say what they do to simplify this line

Reply:
should be much clearer now.